src/mds/MDSContext.h

```diff
@@ -136,10 +138,34 @@ class C_IO_Wrapper : public MDSIOContext
     assert(wrapped != NULL);
   }
   virtual void finish(int r) {
+    if (r == -ECANCELED) {
+      delete wrapped;
+      return;
+    }
     wrapped->complete(r);
   }
 };
 
+/**
+ * Like C_OnFinisher but cancellable, in which case we do
+ * cleanup manually instead of handling it to the finisher.
+ */
+class MDSIOFinisher: public Context {
+  C_IO_Wrapper *con;
+  Finisher *fin;
+public:
+  MDSIOFinisher(C_IO_Wrapper *c, Finisher *f) : con(c), fin(f) {
+    assert(fin != NULL);
+    assert(con != NULL);
+  }
+  void finish(int r) {
```

Comment:
Wouldn't it be simpler to just change C_IO_Wrapper to have a destructor that deletes the wrapped context?  That way the caller doesn't have to know about the magic ECANCELLED behaviour, he can just delete a context if he never wants to call it back